internal/handlers/common/sort.go

```diff
@@ -14,10 +14,160 @@
+		return nil
+	}
+
+	if sort.Len() > 32 {
+		return lazyerrors.Errorf("maximum sort keys exceeded: %v", sort.Len())
+	}
+
+	sortFuncs := make([]sortFunc, len(sort.Keys()))
+	for i, sortKey := range sort.Keys() {
+		sortField := must.NotFail(sort.Get(sortKey))
+
+		sortType, err := getSortType(sortField)
+		if err != nil {
+			return NewErrorMsg(ErrIllegalKey, fmt.Sprintf("%v: %v: %#v", err, sortKey, sortField))
+		}
+
+		sortFuncs[i] = lessFunc(sortKey, sortType)
+	}
+
+	sorter := &docsSorter{docs: docs, sorts: sortFuncs}
+	sorter.Sort(docs)
+
 	return nil
 }
+
+// lessFunc takes sort key and type and returns compare function which
+// compares selected key of 2 documents
+func lessFunc(sortKey string, sortType sortType) func(a, b *types.Document) bool {
+	return func(a, b *types.Document) bool {
+		aField, err := a.Get(sortKey)
+		if err != nil {
+			return false
+		}
+
+		bField, err := b.Get(sortKey)
+		if err != nil {
+			return false
+		}
+
+		switch aField := aField.(type) {
+		case string:
+			bField, err := AssertType[string](bField)
+			if err != nil {
+				return false
+			}
+
+			return strings.Compare(aField, bField) == -1
+		default:
+			result := compareScalars(aField, bField)
+			return matchSortResult(sortType, result)
+		}
+	}
+}
+
+type sortFunc func(a, b *types.Document) bool
+
+type docsSorter struct {
+	docs  []*types.Document
+	sorts []sortFunc
+}
+
+func (ds *docsSorter) Sort(docs []*types.Document) {
+	ds.docs = docs
+	sort.Sort(ds)
+}
+
+func (ds *docsSorter) Len() int {
+	return len(ds.docs)
+}
+
+func (ds *docsSorter) Swap(i, j int) {
+	ds.docs[i], ds.docs[j] = ds.docs[j], ds.docs[i]
+}
+
+func (ds *docsSorter) Less(i, j int) bool {
+	p, q := ds.docs[i], ds.docs[j]
+	// Try all but the last comparison.
+	var k int
+	for k = 0; k < len(ds.sorts)-1; k++ {
+		sortFunc := ds.sorts[k]
+		switch {
+		case sortFunc(p, q):
+			// p < q, so we have a decision.
+			return true
+		case sortFunc(q, p):
+			// p > q, so we have a decision.
+			return false
+		}
+		// p == q; try the next comparison.
+	}
+	// All comparisons to here said "equal", so just return whatever
+	// the final comparison reports.
+	return ds.sorts[k](p, q)
+}
+
+// getSortType determines sortType from input sort value.
+func getSortType(value any) (sortType, error) {
+	sortValue, err := AssertInt64Value("$sort", value)
+	if err != nil {
+		return 0, err
+	}
+
+	switch sortValue {
+	case 1:
+		return ascending, nil
+	case -1:
+		return descending, nil
+	default:
+		return 0, lazyerrors.New(`Illegal key in $sort specification`)
+	}
+}
+
+// matchSortResult matching sort type and compare result.
+func matchSortResult(sort sortType, result compareResult) bool {
+	cmp := false
+	switch result {
+	case less:
+		switch sort {
+		case ascending:
+			cmp = true
+		case descending:
+			cmp = false
+		}
+	case greater:
+		switch sort {
+		case ascending:
+			cmp = false
+		case descending:
+			cmp = true
+		}
+	case notEqual, equal:
+		return false // ???
```

Comment:
what meant be `???` here?

Reply:
removed this comment. it marked notEqual situation but it seems that it's ok to return false.